demos/test-app/src/auth.ts

```diff
@@ -0,0 +1,123 @@
+import type { SignIdentity, Signature } from "@dfinity/agent";
+import {
+  Delegation,
+  DelegationChain,
+  DelegationIdentity,
+} from "@dfinity/identity";
+import { Principal } from "@dfinity/principal";
+
+// The type of response from II as per the spec
+interface AuthResponseSuccess {
+  kind: "authorize-client-success";
+  delegations: {
+    delegation: {
+      pubkey: Uint8Array;
+      expiration: bigint;
+      targets?: Principal[];
+    };
+    signature: Uint8Array;
+  }[];
+  userPublicKey: Uint8Array;
+}
+
+// Perform a sign in to II using parameters set in this app
+export const authWithII = async ({
+  url: url_,
+  maxTimeToLive,
+  derivationOrigin,
+  sessionIdentity,
+}: {
+  url: string;
+  maxTimeToLive?: bigint;
+  derivationOrigin?: string;
+  sessionIdentity: SignIdentity;
+}): Promise<DelegationIdentity> => {
+  // Figure out the II URL to use
+  const iiUrl = new URL(url_);
+  iiUrl.hash = "#authorize";
+
+  // Open an II window and kickstart the flow
+  const win = window.open(iiUrl, "ii-window");
+  if (win === null) {
+    throw new Error(`Could not open window for '${iiUrl}'`);
+  }
+
+  // Wait for II to say it's ready
+  const evnt = await new Promise<MessageEvent>((resolve) => {
+    const readyHandler = (e: MessageEvent) => {
```

Comment:
1. if "message" is never emitted, the listener is never removed

2. should all these listener be provided as return value to the caller that way if the caller destroy itself before this functions ends, the caller can destroy those remaining listeners

Reply:
> if "message" is never emitted, the listener is never removed

right, though if message is never emitted there's nothing else we can do but wait. Would you prefer to have a timeout or something that just errors out after a minute?

> should all these listener be provided as return value to the caller that way if the caller destroy itself before this functions ends, the caller can destroy those remaining listeners

I think that's overkill for a test helper, but I'll keep it in mind if I extract this into a lib. That being said, if the caller decide to destroy itself early it wouldn't have access to the return value yet; IIUC I'd need to return an object that contains both a clean up function and the promise

Reply:
> Would you prefer to have a timeout or something that just errors out after a minute?

Comparing to what I did for Papyrs, looks lik I did something like that to observe if the window/popup got closed.

```
private observeCloseTab() {
    this.closeTabInterval = setInterval(() => {
      if (!this.tab?.closed) {
        return;
      }

      this.clearCloseTabInterval();

      this.throwError('User interrupted sign in.');
    }, 500);
  }
```

I did not removed the listeners because my solution was implemented in a Stencil web components, so destroying the components destroyed the listeners. That's why I would implement the "overkill" so that if the consumer destroy itself, it can destroy the listener.

Generally speaking, keeping listener opens if often a root cause of issues. Don't have statistics but, happens more than once I would guess.

That said, I understand this is just a demo. So feel free to ignore my comment.